QL: Optimize Like/Rlike all

costin opened the pull request:
Replace common Like and RLike queries that match all characters with
IsNotNull (exists) queries

Fix #62585

costin committed:
QL: Optimize Like/Rlike all

Replace common Like and RLike queries that match all characters with
IsNotNull (exists) queries

Fix #62585

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/EQL)

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

matriv commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/regex/LikePattern.java:
```diff
@@ -48,6 +48,16 @@ public String asJavaRegex() {
+    @Override
+    public boolean matchesAll() {
+        for (int i = 0 ; i < pattern.length(); i++) {
+            if (pattern.charAt(i) != '*') {
```
Isn't the wildcard character a `%`? To my understanding the replacement of `%` -> `*` happens later here: https://github.com/elastic/elasticsearch/pull/62682/files#diff-cb6f05d0222f3291ea3df5d33525d6f6R80

costin replied:
You are correct. Fixed this.

matriv reviewed (commented):
Left a comment about the Like/LikePattern.

costin committed:
Fix query translation

costin commented:
I've added more tests to check the query translation and it turns out `IsNull`/`IsNotNull` always defaulted to scripts. So I've fixed the translation rules and added moved the wildcard pattern rule in the optimizer so that according to the comments in #62651:

1. it only gets applied once
2. it ignores folding and works only on literal strings 

It's worth revisiting some of the translations since this might not be singular case.

matriv commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/regex/LikePattern.java:
```diff
@@ -48,6 +48,16 @@ public String asJavaRegex() {
+    @Override
+    public boolean matchesAll() {
+        for (int i = 0 ; i < pattern.length(); i++) {
+            if (pattern.charAt(i) != '%') {
```
I think that now the "problem" is reversed since the ReplaceWildcards takes place before this rule, but the test seems to work: https://github.com/elastic/elasticsearch/pull/62682/files#diff-6dd3fd1d0809ee5978cdccbc6bb7d7e1R17 ? In any case I think we should have a comment mentioning the order of the rules are applied and the choice of % or * in this rule.

costin replied:
`LikePattern` has several styles. In its constructor it accepts the `%` and `_` characters instead of `*` and `?`. The conversion is done internally and accessible.
The check above works on the raw pattern - I'll follow-up with a comment.

matriv commented:
Relates to #62760

costin committed:
Change pattern detection to Lucene automaton

costin commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/62585

costin committed:
Add left-over test

astefan reviewed (approved):
LGTM

matriv reviewed (commented):
LGTM, nice approach with the new checks.
